File: Cs/Solution/SubarraySum.cs
using System;
using System.Collections.Generic;

namespace Solution
{
	public partial class Solution
	{
		public int SubarraySum(int[] nums) {
			int n = nums.Length;
			int[] prefixSum = new int[n + 1];
			for (int i = 0; i < n; i++)
			{
				prefixSum[i + 1] = prefixSum[i] + nums[i];
			}
			int totalSum = 0;
			for (int i = 0; i < n; i++)
			{
				int start = Math.Max(0, i - nums[i]);
				int subarraySum = prefixSum[i + 1] - prefixSum[start];
				totalSum += subarraySum;
			}
			return totalSum;
		}
	}
}

File: Ruby/solution/subarray_sum.rb
# @param {Integer[]} nums
# @return {Integer}
def subarray_sum(nums)
  prefix_sum = [nums[0]]
  for i in 1...nums.size
    prefix_sum[i] = nums[i] + prefix_sum[i - 1]
  end
  prefix_sum.unshift(0)
  sum = 0
  for i in 0...nums.size
    start = [0, i - nums[i]].max
    sum += prefix_sum[i + 1] - prefix_sum[start]
  end
  sum
end
